### PR TITLE
Improve session handling

### PR DESCRIPTION
**Pull request by cbroglie**
This adds the following functionality:
- Make the maximum number of sessions configurable
- Ensure we never open more sessions than the max we specify
- Timeout instead of (potentially) blocking forever waiting on a session
- Close idle sessions

These changes were tested using SoftHSM2.

### CODE REVIEW

File: dsa.go [hunk 1]
```diff
@@ -124,7 +125,7 @@ func GenerateDSAKeyPairOnSession(session pkcs11.SessionHandle, slot uint, id []b
 		pkcs11.NewAttribute(pkcs11.CKA_ID, id),
 	}
 	mech := []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_DSA_KEY_PAIR_GEN, nil)}
-	pubHandle, privHandle, err := libHandle.GenerateKeyPair(session,
+	pubHandle, privHandle, err := libHandle.GenerateKeyPair(session.Handle,
```
Comment (mahrud) [inline dsa.go:128]: Is there a logic behind passing `PKCS11Session` to some functions and `SessionHandle` to others?

File: ecdsa.go [hunk 2]
```diff
@@ -196,15 +197,15 @@ func unmarshalEcPoint(b []byte, c elliptic.Curve) (x *big.Int, y *big.Int, err e
...
+func exportECDSAPublicKey(session *PKCS11Session, pubHandle pkcs11.ObjectHandle) (crypto.PublicKey, error) {
 	var err error
 	var attributes []*pkcs11.Attribute
 	var pub ecdsa.PublicKey
 	template := []*pkcs11.Attribute{
 		pkcs11.NewAttribute(pkcs11.CKA_ECDSA_PARAMS, nil),
 		pkcs11.NewAttribute(pkcs11.CKA_EC_POINT, nil),
 	}
-	if attributes, err = libHandle.GetAttributeValue(session, pubHandle, template); err != nil {
+	if attributes, err = libHandle.GetAttributeValue(session.Handle, pubHandle, template); err != nil {
```
Comment (mahrud) [inline ecdsa.go:208]: Adding to the previous comment, I think having two different "handles" (one `libHandle` and one `session.Handle`) is a little confusing.

File: sessions.go [hunk 3]
```diff
@@ -22,58 +22,84 @@
...
+
+	return f(session.(*PKCS11Session))
 }
 
 // Create the session pool for a given slot if it does not exist
 // already.
-func setupSessions(slot uint, max int) error {
+func setupSessions(slot uint) error {
 	sessionPoolMutex.Lock()
-	defer sessionPoolMutex.Unlock()
```
Comment (mahrud) [inline sessions.go:71]: Is there a benefit to putting this in the end?

File: crypto11.go [hunk 4]
```diff
@@ -202,16 +214,28 @@ func Configure(config *PKCS11Config) (*pkcs11.Ctx, error) {
...
-	if err = setupSessions(defaultSlot, 0); err != nil {
+	defaultSlot = slot
+	flags = token.Flags
+
+	maxSessions = config.MaxSessions
+	if maxSessions == 0 {
+		maxSessions = DefaultMaxSessions
+	}
+	if token.MaxRwSessionCount > 0 && uint(maxSessions) > token.MaxRwSessionCount {
+		return nil, fmt.Errorf("crypto11: provided max sessions value (%d) exceeds max value the token supports (%d)", maxSessions, token.MaxRwSessionCount)
```
Comment (mahrud) [inline crypto11.go:230]: Use `log.Fatalf` instead since `log` is already imported.

File: sessions.go [hunk 5]
```diff
@@ -22,58 +22,84 @@
...
+		return fmt.Errorf("crypto11: no session for slot %d", slot)
 	}
-	defer func() {
-		// TODO better would be to close the session if the pool is full
-		sessionPool <- session
-	}()
-	return f(session)
+
+	ctx, cancel := context.WithTimeout(context.Background(), newSessionTimeout)
+	defer cancel()
```
Comment (mahrud) [inline sessions.go:77]: Is there a probability that this would cut off a session that is active?

**Review by mahrud — CHANGES_REQUESTED** [review]
Also works with YubiKey when `MaxSessions=1`.

File: dsa.go (hunk 1 above)
Comment (cbroglie) [reply dsa.go:128]: All of the functions in crypto11 should operate on `*PKCS11Session`, whereas all of the functions in pkcs11 accept a `pkcs11.SessionHandle`.

File: ecdsa.go (hunk 2 above)
Comment (cbroglie) [reply ecdsa.go:208]: That's the nature of pkcs11 - there are many handles. I'm just maintaining the existing conventions, see `PKCS11Object` for another example:
```go
// PKCS11Object contains a reference to a loaded PKCS#11 object.
type PKCS11Object struct {
	// The PKCS#11 object handle.
	Handle pkcs11.ObjectHandle

	// The PKCS#11 slot number.
	//
	// This is used internally to find a session handle that can
	// access this object.
	Slot uint
}
```

File: sessions.go (hunk 3 above)
Comment (cbroglie) [reply sessions.go:71]: There is a very minor (negligible here) performance impact of using `defer`, so for simple functions without any early returns I've developed the habit of not using it.

File: crypto11.go (hunk 4 above)
Comment (cbroglie) [reply crypto11.go:230]: I'm merely returning an error here, not trying to make the application exit (which is what `log.Fatal` will do). Handling the error is up to the calling application.

File: sessions.go (hunk 5 above)
Comment (cbroglie) [reply sessions.go:77]: Good question, but no, the context is only used as a timeout inside `Get`.

**Comment by mahrud** [comment]
Confirmed that also works with the Thales nShield HSM.

File: crypto11.go (hunk 4 above)
Comment (mahrud) [reply crypto11.go:230]: I understand, but if you look above, this same function is logging errors in addition to returning errors. I think it's a good idea given that this function handles hardware. What do you think?

File: sessions.go (hunk 5 above)
Comment (mahrud) [reply sessions.go:77]: Would still like to test this in practice when we can.

File: crypto11.go (hunk 4 above)
Comment (optnfast) [reply crypto11.go:230]: Current behavior (i.e.without this PR) is a bit inconsistent; I've created https://github.com/ThalesIgnite/crypto11/issues/4 as a placeholder for doing a cleanup at some later date.

**Comment by optnfast** [comment]
Thanks for the PR!

**Pull request merged**